src/iceberg/expression/decimal.h

```diff
@@ -0,0 +1,274 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#pragma once
+
+/// \file iceberg/expression/decimal.h
+/// \brief 128-bit fixed-point decimal numbers.
+/// Adapted from Apache Arrow with only Decimal128 support.
+/// https://github.com/apache/arrow/blob/main/cpp/src/arrow/util/decimal.h
+
+#include <array>
+#include <cstdint>
+#include <iosfwd>
+#include <string>
+#include <string_view>
+#include <type_traits>
+
+#include "iceberg/iceberg_export.h"
+#include "iceberg/result.h"
+#include "iceberg/util/macros.h"
+#include "iceberg/util/port.h"
+
+namespace iceberg {
+
+/// \brief Represents 128-bit fixed-point decimal numbers.
+/// The max decimal precision that can be safely represented is
+/// 38 significant digits.
+class ICEBERG_EXPORT Decimal {
+ public:
+  static constexpr int32_t kBitWidth = 128;
+  static constexpr int32_t kByteWidth = kBitWidth / 8;
+  static constexpr int32_t kMaxPrecision = 38;
+  static constexpr int32_t kMaxScale = 38;
+
+  /// \brief Default constructor initializes to zero.
+  constexpr Decimal() noexcept : data_{0, 0} {}
+
+  /// \brief Construct a Decimal from an array of 2 64-bit integers.
+  explicit Decimal(std::array<uint64_t, 2> data) noexcept : data_(std::move(data)) {}
```

Comment:
We are missing a ctor of `Decimal(const uint8_t*, size_t)`. This is required to restore encoded decimal values from binary data.

Reply:
There is a `FromBigEndian` that could serve the same purpose. I initially avoided the ctor because it may throw an exception on invalid input, maybe the `FromBigEndian` is good enough?